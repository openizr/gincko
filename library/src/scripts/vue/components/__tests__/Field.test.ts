/**
 * Copyright (c) Matthieu Jabbour. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import Vue from 'vue';
import { mount } from '@vue/test-utils';
import Field from 'scripts/vue/components/Field.vue';

jest.mock('sonar-ui/vue', () => {
  const Component = {
    render(createElement: Json): Json {
      const self = this as Json;
      return createElement('div', {
        attrs: {
          id: self.id,
          'data-readonly': self.readonly,
        },
        on: {
          click(): void {
            self.$emit('click');
          },
          keyDown(): void {
            self.$emit('focus');
          },
        },
      });
    },
  };

  const markdown = (value: string): string => value;
  const buildClass = (...values: string[]): string => values.join(' ');
  const UIRadio = Vue.extend(Component);
  const UIButton = Vue.extend(Component);
  const UIDropdown = Vue.extend(Component);
  const UICheckbox = Vue.extend(Component);
  const UITextfield = Vue.extend(Component);
  const UIFileUploader = Vue.extend(Component);

  return {
    markdown,
    buildClass,
    UIRadio,
    UIButton,
    UIDropdown,
    UICheckbox,
    UITextfield,
    UIFileUploader,
  };
});

describe('vue/components/Field', () => {
  const onUserAction = jest.fn();
  const customComponents = {};

  beforeEach(() => {
    jest.clearAllMocks();
  });

  test('unknown field type', () => {
    const wrapper = mount(Field, {
      propsData: {
        id: 'Message',
        type: 'Unknown',
        status: 'initial',
        options: {},
        customComponents,
      },
      listeners: {
        userAction: onUserAction,
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('Message with empty label', () => {
    const wrapper = mount(Field, {
      propsData: {
        id: 'Message',
        type: 'Message',
        status: 'initial',
        options: {},
      },
      listeners: {
        userAction: onUserAction,
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('Message with label', () => {
    const wrapper = mount(Field, {
      propsData: {
        id: 'Message',
        type: 'Message',
        status: 'initial',
        label: 'Test {{value}}',
        options: { formValues: { value: 'test' } },
        customComponents,
      },
      listeners: {
        userAction: onUserAction,
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('Button', async () => {
    const wrapper = mount(Field, {
      propsData: {
        id: 'Button',
        active: true,
        type: 'Button',
        status: 'initial',
        label: 'Test',
        options: {},
        customComponents,
      },
      listeners: {
        userAction: onUserAction,
      },
    });
    await wrapper.trigger('click');
    expect(wrapper.html()).toMatchSnapshot();
    expect(onUserAction).toHaveBeenCalledTimes(1);
  });

  test('Textfield - active then inactive step', async () => {
    const wrapper = mount(Field, {
      propsData: {
        id: 'Textfield',
        active: true,
        type: 'Textfield',
        status: 'initial',
        label: 'Test',
        options: {},
        customComponents,
      },
      listeners: {
        userAction: onUserAction,
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
    wrapper.setProps({ active: false });
    await wrapper.vm.$nextTick();
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('Textfield - active step, readonly option is `true`', () => {
    const wrapper = mount(Field, {
      propsData: {
        id: 'Textfield',
        active: true,
        type: 'Textfield',
        status: 'initial',
        label: 'Test',
        options: { readonly: true },
        customComponents,
      },
      listeners: {
        userAction: onUserAction,
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('Textfield - with onFocus method, inactive step', async () => {
    const onFocus = jest.fn();
    const wrapper = mount(Field, {
      propsData: {
        id: 'Textfield',
        active: false,
        type: 'Textfield',
        status: 'initial',
        label: 'Test',
        options: { onFocus },
        customComponents,
      },
      listeners: {
        userAction: onUserAction,
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
    expect(onFocus).toHaveBeenCalledTimes(0);
    await wrapper.trigger('keyDown');
    await wrapper.vm.$nextTick();
    expect(wrapper.html()).toMatchSnapshot();
    expect(onFocus).toHaveBeenCalledTimes(1);
  });

  test('Textfield - with no onFocus method', async () => {
    const wrapper = mount(Field, {
      propsData: {
        id: 'Textfield',
        active: false,
        type: 'Textfield',
        status: 'initial',
        label: 'Test',
        options: {},
        customComponents,
      },
      listeners: {
        userAction: onUserAction,
      },
    });
    await wrapper.trigger('keyDown');
    await wrapper.vm.$nextTick();
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('FileUploader', () => {
    const wrapper = mount(Field, {
      propsData: {
        id: 'FileUploader',
        active: true,
        type: 'FileUploader',
        status: 'initial',
        label: 'Test',
        options: {},
        customComponents,
      },
      listeners: {
        userAction: onUserAction,
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('Dropdown', () => {
    const wrapper = mount(Field, {
      propsData: {
        id: 'Dropdown',
        active: true,
        type: 'Dropdown',
        status: 'initial',
        label: 'Test',
        options: {},
        customComponents,
      },
      listeners: {
        userAction: onUserAction,
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('Checkbox', () => {
    const wrapper = mount(Field, {
      propsData: {
        id: 'Checkbox',
        active: true,
        type: 'Checkbox',
        status: 'initial',
        label: 'Test',
        options: {},
        customComponents,
      },
      listeners: {
        userAction: onUserAction,
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('Radio', () => {
    const wrapper = mount(Field, {
      propsData: {
        id: 'Radio',
        active: true,
        type: 'Radio',
        status: 'initial',
        label: 'Test',
        options: {},
        customComponents,
      },
      listeners: {
        userAction: onUserAction,
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
});
