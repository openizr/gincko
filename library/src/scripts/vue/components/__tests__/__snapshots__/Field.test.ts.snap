// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`vue/components/Field Button 1`] = `"<div id=\\"Button\\" label=\\"Test\\" modifiers=\\"initial  \\"></div>"`;

exports[`vue/components/Field Checkbox 1`] = `"<div id=\\"Checkbox\\" name=\\"Checkbox\\" label=\\"Test\\" options=\\"[object Object],[object Object]\\" modifiers=\\"initial \\"></div>"`;

exports[`vue/components/Field Dropdown 1`] = `"<div id=\\"Dropdown\\" name=\\"Dropdown\\" label=\\"Test\\" options=\\"[object Object],[object Object]\\" modifiers=\\"initial \\"></div>"`;

exports[`vue/components/Field FileUploader 1`] = `"<div id=\\"FileUploader\\" name=\\"FileUploader\\" label=\\"Test\\" modifiers=\\"initial \\"></div>"`;

exports[`vue/components/Field Message with empty label 1`] = `"<section id=\\"Message\\" class=\\"ui-message initial,\\"></section>"`;

exports[`vue/components/Field Message with label 1`] = `"<section id=\\"Message\\" class=\\"ui-message initial,\\">Test test</section>"`;

exports[`vue/components/Field Radio 1`] = `"<div id=\\"Radio\\" name=\\"Radio\\" label=\\"Test\\" options=\\"[object Object],[object Object]\\" modifiers=\\"initial \\"></div>"`;

exports[`vue/components/Field Textarea - active step, readonly option is \`true\` 1`] = `"<div id=\\"Textarea\\" name=\\"Textarea\\" label=\\"Test\\" debouncetimeout=\\"100\\" readonly=\\"readonly\\" modifiers=\\"initial \\"></div>"`;

exports[`vue/components/Field Textarea - active then inactive step 1`] = `"<div id=\\"Textarea\\" name=\\"Textarea\\" label=\\"Test\\" debouncetimeout=\\"100\\" modifiers=\\"initial \\"></div>"`;

exports[`vue/components/Field Textarea - active then inactive step 2`] = `"<div id=\\"Textarea\\" name=\\"Textarea\\" label=\\"Test\\" debouncetimeout=\\"100\\" modifiers=\\"initial \\" readonly=\\"readonly\\"></div>"`;

exports[`vue/components/Field Textarea - with no onFocus method 1`] = `"<div id=\\"Textarea\\" name=\\"Textarea\\" label=\\"Test\\" debouncetimeout=\\"100\\" modifiers=\\"initial \\"></div>"`;

exports[`vue/components/Field Textarea - with onFocus method, inactive step 1`] = `"<div id=\\"Textarea\\" name=\\"Textarea\\" label=\\"Test\\" debouncetimeout=\\"100\\" readonly=\\"readonly\\" modifiers=\\"initial \\"></div>"`;

exports[`vue/components/Field Textarea - with onFocus method, inactive step 2`] = `"<div id=\\"Textarea\\" name=\\"Textarea\\" label=\\"Test\\" debouncetimeout=\\"100\\" modifiers=\\"initial \\"></div>"`;

exports[`vue/components/Field Textfield - active step, readonly option is \`true\` 1`] = `"<div id=\\"Textfield\\" name=\\"Textfield\\" label=\\"Test\\" readonly=\\"readonly\\" debouncetimeout=\\"100\\" modifiers=\\"initial  \\"></div>"`;

exports[`vue/components/Field Textfield - active then inactive step 1`] = `"<div id=\\"Textfield\\" name=\\"Textfield\\" label=\\"Test\\" debouncetimeout=\\"100\\" modifiers=\\"initial  \\"></div>"`;

exports[`vue/components/Field Textfield - active then inactive step 2`] = `"<div id=\\"Textfield\\" name=\\"Textfield\\" label=\\"Test\\" debouncetimeout=\\"100\\" modifiers=\\"initial  \\" readonly=\\"readonly\\"></div>"`;

exports[`vue/components/Field Textfield - with no onFocus method 1`] = `"<div id=\\"Textfield\\" name=\\"Textfield\\" label=\\"Test\\" debouncetimeout=\\"100\\" modifiers=\\"initial  \\"></div>"`;

exports[`vue/components/Field Textfield - with onFocus method, inactive step 1`] = `"<div id=\\"Textfield\\" name=\\"Textfield\\" label=\\"Test\\" readonly=\\"readonly\\" debouncetimeout=\\"100\\" modifiers=\\"initial  \\"></div>"`;

exports[`vue/components/Field Textfield - with onFocus method, inactive step 2`] = `"<div id=\\"Textfield\\" name=\\"Textfield\\" label=\\"Test\\" debouncetimeout=\\"100\\" modifiers=\\"initial  \\"></div>"`;

exports[`vue/components/Field unknown field type 1`] = `""`;
